common/src/lib.rs

```diff
@@ -0,0 +1,163 @@
+#[async_trait]
+pub trait KVStore {
+    /// Creates an empty store with the path to newly create.
+    async fn new(path: &str) -> Result<Self, ()>
+    where
+        Self: Sized;
+    /// Open an existing store with the path given.
+    async fn open(path: &str) -> Result<Self, ()>
+    where
+        Self: Sized;
+    /// Records the current state to the persistent storage.
+    async fn commit_checkpoint(&mut self) -> Result<(), ()>;
+    /// Reverts all the changes made since the last checkpoint.
+    async fn reset_to_latest_checkpoint(&mut self) -> Result<(), ()>;
+    /// Inserts a key-value pair into the store. If exists, it will be overwritten.
+    async fn insert(&mut self, key: Hash256, value: &[u8]) -> Result<(), ()>;
+    /// Remove a key-value pair from the store. If not exists, it will fail.
+    async fn remove(&mut self, key: Hash256) -> Result<(), ()>;
+    /// Retrieves the value associated with the key. If not exists, it will return `None`.
+    async fn get(&self, key: Hash256) -> Result<Option<Vec<u8>>, ()>;
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
+pub struct Header {
+    /// The author of this block.
+    pub author: PublicKey,
+    /// The signature of the previous block.
+    pub prev_signature: Vec<Signature>,
+    /// The hash of the previous block.
+    pub previous_hash: Hash256,
+    /// The height of this block.
+    pub height: u64,
+    /// The timestamp of this block.
+    pub timestamp: u64,
+    /// The Merkle root of transactions.
+    pub tx_merkle_root: Hash256,
+    /// The Merkle root of the state.
+    pub state_merkle_root: Hash256,
+    /// The hash of the set of validator & vote weight for the next block.
+    pub validator_set_hash: Hash256,
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
+pub enum StateTransition {
+    InsertValidator {
+        /// The public key of the validator.
+        public_key: PublicKey,
+        /// The weight of the validator.
+        weight: u64,
+    },
+    RemoveValidator(PublicKey),
+    /// This is accepted only if the signer of the transaction equals to the `delegator`.
+    Delegate {
+        /// The public key of the validator who delegates its voting right.
+        delegator: PublicKey,
+        /// The public key of the validator who is being delegated.
+        delegatee: PublicKey,
+        /// The target height of the block of this transaction, which is for preventing the replay attack.
+        target_height: u64,
+    },
+    /// This is accepted only if the signer of the transaction equals to the `delegator`.
+    Undelegate {
+        /// The public key of the validator who claims its voting right.
+        delegator: PublicKey,
+        /// The target height of the block of this transaction, which is for preventing the replay attack.
+        target_height: u64,
+    },
+    InsertData {
+        key: String,
+        value: String,
+    },
+    RemoveData(String),
+    Noop,
```

Comment:
Why do we need this Noop?

Reply:
For those txes that carry some `data` (refer to `Transaction`) but has nothing to do with the state.
It is useful when we want to deliver some one-time message to the colony chains.